network/gossip/libp2p/libp2pNode_test.go

```diff
@@ -332,75 +329,62 @@ func (l *LibP2PNodeTestSuite) TestCreateStreamTimeoutWithUnresponsiveNode() {
 	// attempt to create a stream from node 1 to node 2 and assert that it fails after timeout
 	grace := 10 * time.Millisecond
 	var err error
-	unittest.AssertReturnsBefore(l.T(),
+	unittest.AssertReturnsBefore(suite.T(),
 		func() {
 			_, err = peers[0].CreateStream(ctx, silentNodeAddress)
 		},
 		DefaultUnicastTimeout+grace)
-	assert.Error(l.T(), err)
+	assert.Error(suite.T(), err)
 }
 
 // TestCreateStreamIsConcurrent tests that CreateStream calls can be made concurrently such that one blocked call
 // does not block another concurrent call.
-func (l *LibP2PNodeTestSuite) TestCreateStreamIsConcurrent() {
+func (suite *LibP2PNodeTestSuite) TestCreateStreamIsConcurrent() {
 
 	// bump up the unicast timeout to a high value
 	unicastTimeout = time.Hour
 
 	// create two regular node
-	goodPeers, goodAddrs := l.CreateNodes(2, nil, false)
-	defer l.StopNodes(goodPeers)
-	require.Len(l.T(), goodAddrs, 2)
+	goodPeers, goodAddrs := suite.CreateNodes(2, nil, false)
+	defer suite.StopNodes(goodPeers)
+	require.Len(suite.T(), goodAddrs, 2)
 
 	// create a silent node which never replies
-	listener, silentNodeAddress := newSilentNode(l.T())
-	defer listener.Close()
-
-	wg := sync.WaitGroup{}
-	wg.Add(1)
-
-	ch := make(chan struct{})
-	// spin off a go routine to create a stream to the unresponsive node
-	go func() {
-		wg.Done()
-		_, _ = goodPeers[0].CreateStream(l.ctx, silentNodeAddress) // this call will block
-		close(ch)
+	listener, silentNodeAddress := newSilentNode(suite.T())
+	defer func() {
+		require.NoError(suite.T(), listener.Close())
 	}()
 
-	// make sure the go routine to create a stream with the unresponsive node actually started
-	unittest.AssertReturnsBefore(l.T(), wg.Wait, 5*time.Millisecond)
-	// make sure the go routine to create a stream did not finish
-	select {
-	case <-time.After(10 * time.Millisecond):
-	case <-ch:
-		assert.Fail(l.T(), "CreateStream attempt to the unresponsive peer did not block")
-	}
+	// creates a stream to unresponsive node and makes sure that the stream creation is blocked
+	streamCreated := unittest.RequireNeverReturnBefore(suite.T(),
+		func() {
+			_, _ = goodPeers[0].CreateStream(suite.ctx, silentNodeAddress) // this call will block
+		},
+		1*time.Second,
```

Comment:
These changes basically simplify some test logics using encapsulation, however, they do not change the logic. The abstractions allowed me to find the root cause: the anonymous function and `RequireReturnBefore` are running on two separate goroutinges, which are not necessarily synchronized by go runtime. So, based on watching the goroutines scheduling, there were cases that `RequireReturnBefore` would execute and terminate before the anonymous function of its interest is executed, and it would fail the test. Solution was simply to increase the timeout. However, I wanted to make sure of the root cause before just adhocly adding some timeout, as it would not solve the falkiness, and would also lasten build longer.

Reply:
does that mean there were times when the CI didn't really kick off the the `CreateStream` go routine within `5ms`?
Good catch and `unittest.RequireNeverReturnBefore` is definitely a better way of doing the wait that what I had.

Reply:
Yes, actually this test would even get flakey on my laptop, and as I logged the calls, there were traces where the causality would be violated depending on the number of pending goroutines.